url888b6fd9b4_jmacglashan_commandsToTasks_tgz-pJ8-commands_tests_Model3TestsJ8/src/commands/model3/mt/em/MTEMModule.java

```diff
@@ -0,0 +1,306 @@
+				 */
+				
+				for(int i = 1; i <= m; i++){
+
+					double sc = singleDistortionCounts.v(new IntTupleHash(i, l, m));
+					for(int j = 0; j <= l; j++){
+						double jc = jointDistortionCounts.v(new IntTupleHash(j, i, l, m));
+						
+						double p = jc > 0. ? jc / sc : 0.;
+						dp.set(p, j, i, l, m);
+						
+					}
+					
+				}
+				
+				
+			}
+		}
+		
+		
+		
+		
+		//reset storage for next EM step
+		jointWordCounts = new HashedAggregator<String>(wordAdditiveConstant);
+		singleWordCounts = new HashedAggregator<String>(wordAdditiveConstant*this.mtModule.getNaturalWords().size());
+		
+		jointDistortionCounts = new HashedAggregator<IntTupleHash>(0);
+		singleDistortionCounts = new HashedAggregator<IntTupleHash>(0);
+		
+		
+		this.numEMIterations++;
+		if(this.numEMIterations >= this.numWithWordParamAlone){
+			this.factorInAlignment = true;
+		}
+		
+		this.gm.emptyCache();
+
+	}
+	
+	
+	
+	protected double expectedVal(TokenedString prodText, TokenedString genText, int i, int j, int m, int l){
+		
+		WordParam wp = this.mtModule.getWp();
+		DistortionParam dp = this.mtModule.getDp();
+		
+		double sum = 0.;
+		for(int u = 0; u <= l; u++){
+			double d = 1.;
+			if(this.factorInAlignment){
+				d = dp.prob(u, i, l, m);
+			}
+			double w = wp.prob(prodText.t(i), genText.t(u));
+			sum += d*w;
+		}
+		
+		double d = 1.;
+		if(this.factorInAlignment){
+			d = dp.prob(j, i, l, m);
+		}
+		double w = wp.prob(prodText.t(i), genText.t(j));
+		
+		double num = d*w;
+		double p = num / sum;
+		
+		if(num == 0. && sum == 0.){
+			p = 0.;
+		}
+		
+		if(Double.isNaN(p)){
+			throw new RuntimeException("Expected Val is NaN.");
+		}
+		
+		return p;
+		
+	}
+	
+	
+	
+	
+	protected static String tokenCombine(String prodWord, String genWord){
+		return prodWord + "+++" + genWord;
+	}
+	
+	
+	public static List<WeightedMTInstance> getVanillaMTDataset(Tokenizer tokenizer, String pathToDataDir, String dataFileExtension){
+		
+		List<WeightedMTInstance> dataset = new ArrayList<WeightedMTInstance>();
+		
+		//get rid of trailing /
+		if(pathToDataDir.charAt(pathToDataDir.length()-1) == '/'){
+			pathToDataDir = pathToDataDir.substring(0, pathToDataDir.length());
+		}
+		
+		File dir = new File(pathToDataDir);
+		final String ext = new String(dataFileExtension);
+		
+		FilenameFilter filter = new FilenameFilter() {
+			public boolean accept(File dir, String name) {
+				if(name.endsWith(ext)){
+					return true;
+				}
+				return false;
+			}
+		};
+		String[] children = dir.list(filter);
+		for(int i = 0; i < children.length; i++){
+			String path = pathToDataDir + "/" + children[i];
+			dataset.add(getInstanceFromFile(tokenizer, path));
+		}
+		
+		return dataset;
+	}
+	
+	public static WeightedMTInstance getInstanceFromFile(Tokenizer tokenizer, String pathToFile){
+		
+		WeightedMTInstance d = null;
+		
+		try {
+			BufferedReader in = new BufferedReader(new FileReader(pathToFile));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *in*. The resource is referenced at lines 289, 290, 291. The resource is closed at line 296. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).